lms/views/basic_lti_launch.py

```diff
@@ -210,6 +210,9 @@ def unconfigured_basic_lti_launch(self):
                     "google_developer_key"
                 ],
                 "lmsUrl": self.context.lms_url,
+                "registeredLmsUrl": self.request.find_service(name="ai_getter").lms_url(
+                    oauth_consumer_key
+                ),
```

Comment:
Isn't `registeredLmsUrl` the same value as the `lmsUrl` variable above it?

Reply:
No, lmsUrl is the lmsUrl from the client's point of view (the lms url you currently have open in the browser), registeredLmsUrl is the url from the database that was provided when registering a new lms application with Hypothesis.

Reply:
To expand on Hannah's comment above: When the user clicks the "Authorize" button in the client, that opens the `canvas_api.authorize` route's URL in a popup window. That URL in turn redirects to the Canvas instance's OAuth 2 authorization endpoint. The `canvas_api.authorize` route uses the application instance's LMS URL from our database. The `lmsUrl` variable here on the other hand is the `custom_canvas_api_domain` variable from the LTI launch, if present, with a fallback to the application instance's LMS URL from the database. The `lmsUrl` frontend config key is currently used by the Google Picker, which needs to know the origin of the top-level frame in the window, if the LTI launch happens inside an iframe.

If `lmsUrl` and `registeredLmsUrl` are different here, that probably does indicate a misconfiguration of the LTI installation though, one that ideally we would have caught at the point when the LTI installation is registered.

Reply:
It's called `lms_url` in `ApplicationInstance` but has been renamed to `registeredLmsUrl` here (understandably, since something else called `lmsUrl` is already here).

Suggest maybe naming it `applicationInstanceLmsUrl` instead?

Something that avoids introducing a second name for the same thing ("application instance" vs "registration" / "registered"). "Register" / "registration" is not currently used in the code to refer to this concept. (It does appear in a UI string: "Register here..." but that's all.) 

The existing `lmsUrl` name (the `custom_canvas_api_domain` variable from the LTI launch, if present, with a fallback to the application instance's LMS URL from the database) might not be a great name itself either. There might be a better name for that, that better captures what it is or what it's used for and distinguishes it from `applicationInstanceLmsUrl`.

Anyway, I'd like us to consider the naming here, as that's important for code clarity, and having both `lmsUrl` and `registeredLmsUrl` in the same place isn't super clear.

Reply:
One approach might be to name them `applicationInstanceLmsUrl` (or just `lmsUrl`) and `customCanvasApiDomain` and allow `customCanvasApiDomain`  to be null rather than falling back. Then put the  `customCanvasApiDomain || applicationInstanceLmsUrl` logic in the JavaScript code where `customCanvasApiDomain` is used instead. Not sure if that will work out well or not, but it does avoid the naming problem or the introduction of confusing names

Reply:
> . One approach might be to name them applicationInstanceLmsUrl (or just lmsUrl) and customCanvasApiDomain and allow customCanvasApiDomain to be null rather than falling back. Then put the customCanvasApiDomain || applicationInstanceLmsUrl logic in the JavaScript code where customCanvasApiDomain is used instead. Not sure if that will work out well or not, but it does avoid the naming problem or the introduction of confusing names.

I generally like this approach although it requires a bit of refactoring.